### DISCUSSION TITLE
When is expected a stable version of WinForm core Designer ?

### DISCUSSION THREAD

**Discussion by IrakliLomidze** [dotnet/winforms#5220] [Q&A] [answered]
When is expected a stable version of WinForm core Designer?

.NET 5 Designer was not updated for while.

**Comment by merriemcgaw** [comment, marked as answer]
We've been working hard under the hood to improve our Core designer and bring it up to parity. We are getting close, and 16.10 saw a lot of performance improvements especially with interacting directly with controls on the form, and adding the MenuStrip, ToolStrip and StatusStrip in-situ editors so you can now manage  your menus the way you always have.

When will we be done? I don't have a solid timeline for you. Although, aside from a couple of features, including Accessibility support, Tab Order Overlay and some data binding scenarios you should find most of what is  used on a daily basis. I encourage everyone to install the 16.10 Visual Studio and start playing with the designer. We will be able to fully stabilize and remove the "preview" checkbox only if we've gotten feedback from customers. As we uncover new scenarios that are needed, or old ones that we still need to bring back, your feedback through the VS feedback tool is critical to us so we can prioritize them.

I know it's been a long journey, but we're getting to the finish line. Converting to a 2 process solution, where VS in the .NET Framework can interact with your .NET 5/6 app displaying in the designer has required us to rearchitect the entire way a developer interacts within the designer. As we've done that work we've been working through a system largely stable since VS2005. It's been an exciting journey and we are glad to have everyone on it with us.

**Reply by kirsan31** [reply]
@merriemcgaw and again no word about support of custom user controls. 😥

**Reply by merriemcgaw** [reply]
Can you tell me what isn't working in your current UserControls? We have everything in place for standard custom controls. If you've got User Controls in your projects and they're correctly referenced they should be perfectly usable. If you're referring to Control Libraries it depends on the design time experience you want. If the PropertyGrid is enough you should be good to go. If you need more direct interaction with VS, you will need an upcoming SDK and possibly to work with our new model for cross process communication and interaction between the control on the designer and anything happening in Visual Studio.

 What we're doing currently is cleaning up the SDK for third party control developers who need to interact with Visual Studio in their design time experience. We're working directly with major control vendors to make sure they're able to use the new model for interacting with VS at design time. We're at the clean up stage for the SDK so that we know we're exposing the right surface area and not too much or too little. I would expect a public release of this SDK sometime early in the 17.x timeframe. 

If we've missed specific scenarios aside from complex controls with a lot of design time VS interaction, I'd love to know them ASAP. We would want to get that cleaned up for 17.0 if at all possible. 

Does this help?

**Reply by kirsan31** [reply]
> If you're referring to Control Libraries it depends on the design time experience you want.

Yes, exactly this point.

> If the PropertyGrid is enough you should be good to go.

Sadly, but in 16.11 p2 it's still not the case :( The very good example here is abandoned [MS chart](https://github.com/dotnet/winforms-datavisualization). New designer **not see this control at all**, and will silently corrupt source file after any change in it.  We briefly  discuss this some long time ago, and I see no positive changes here :(

Old designer:
![image](https://user-images.githubusercontent.com/17767561/125157545-467d3a00-e174-11eb-8e2d-843b9cc593ef.png)

New designer:

https://user-images.githubusercontent.com/17767561/125157549-5006a200-e174-11eb-9983-3aa2b62d8c90.mp4

So, we strongly stick to old net framework designer here through `<TargetFrameworks>net472;net5.0-windows</TargetFrameworks>` technique. And I can tell you that this is nightmare. Because the code grows over time with:
```cs
#if !NETFRAMEWORK
#else
#endif
```

Our thots on this:
- We are afraid of using old designer since this technique may become unavailable in the future and due to `defines` point above.
- We even considered to move from `winforms-datavisualization`, but all WinForms chart controls (free) on the market not even close in terms of functionality.
- We cannot move **all** chart logic out of the `*.Designer.cs` due to layout problems and this is not WinForms was designed for.  
  > If the PropertyGrid is enough you should be good to go.
  
  This would be a solution (partially) to the problem - ability to layout and using basic property is must have.
- We still **absolutely no Idea** how much effort is needed to adapt `winforms-datavisualization` control to new designer.

**Reply by merriemcgaw** [reply]
@kirsan31 I'm talking with the team and taking your feedback about the Chart control. We know that you and others are dependent on it. I am going to suggest that we should perhaps open the repo enough to take PRs to create designer for the Chart Control. You may be able to get a head start using the instructions for custom control designers for .NET Framework if all you need is the PropertyGrid support. That effort should be minimal in comparison with creating a complex design time experience interacting with both the Core control and the Framework Visual Studio. The SDK to support that scenario is currently being refined but should be available in a preview version with preview documentation in the VS 2022 timeframe.

**Reply by kirsan31** [reply]
@merriemcgaw thank you for answer.

> I'm talking with the team and taking your feedback about the Chart control. We know that you and others are dependent on it. I am going to suggest that we should perhaps open the repo enough to take PRs to create designer for the Chart Control. You may be able to get a head start using **the instructions for custom control designers for .NET Framework** if all you need is the PropertyGrid support.

Any timeframe when these instructions will be available?
P.s. Am I understand right, that your mean _.NET_ instead of _.NET Framework_ in phrase above?

**Comment by IrakliLomidze** [comment]
merriemcgaw

The timeline for finalizing WinForm Core tooling is extremely important at least in my case.
I'm going to ship a new version of my desktop application this fall. But still not sure about stay on .NET Framework or will be safe to move to .NET 6.

Can you clarify a bit, expected roadmap of WinForm Designer?

**Reply by merriemcgaw** [reply]
From a runtime perspective I'd recommend that you do move to .NET 6 because of all the performance enhancements and accessibility improvements we've been making since going open source. You can use the workaround @kirsan31 mentions above for now to utilize the Framework Designer, the only drawback at the moment is that you wouldn't be able to visualize things that are new in .NET that weren't in .NET Framework 4.8, like the new default font property for example. 

For the most part, we're close to getting the core pieces of the designer in place. I don't have exact dates for you, but I can tell you that these are the things we're working on for the next couple of releases:

1. Tab Order Overlay to help set the tab order in your application
2. SDK for creating design time experience on custom controls. Note that most controls where  PropertyGrid support is all that is needed should be able to work the way they used to. It's only when we're working with complex design experience (like with 3rd party controls) that you have to use our new SDK, otherwise if it's simple proceed as before with Custom Controls
3. Performance performance performance. Moving to this multi-process environment was costly to the  user experience so we're focusing on it.
4. VB support in the new Property Pages, and addition of our new properties to the Property Pages in VS 2022.
5. Accessibility support.
6. Responding to customer feedback

I hope this helps you make a call to use .NET 6. Porting applications is going to get more difficult from a code perspective the more WinForms .NET and WinForms .NET Framework diverge. Given we've got a workaround that lets you use the Framework Designer while also building for .NET, I wouldn't let the tooling stop you at this point from porting.

**Comment by cadilhac** [comment]
And what is the future for licensing winforms components? As a component vendor, I'm stuck with .net framework 4.x and can't migrate to .net 5 or 6 and propose it to my customers because the old mechanism (licenses.licx file / using a 3rd party product like Crypto Licensing) does not work anymore. Is it in the roadmap? When can we expect a workable solution and some docs to properly license components?

**Reply by kirsan31** [reply]
There are big problems with the new designer. Obviously it won't even be fully ready for the studio's 2022 release. Also, as far as I understand, due to same problem, we did not receive the promised revision of HDPI in .Net6 (moved to .Net7) :(

**Reply by merriemcgaw** [reply]
In order to make the new designer work with Core projects - since Visual Studio is a .NET process we've had to completely redo the inner workings of the designer. That's taken more time than we'd like but we're getting closer with every preview release. By VS 2022 release we intend to have docs and how-to's for creating complex custom controls that have to communicate with both .NET and .NET Framework. We are aiming to be in a place where our SDK is solid enough that everyone can start depending on it. Other than that, most features are in place, and if you're running into blocking issues please by all means file the feedback in VS so we can take a look at this.

We've also put some data binding capabilities to account for not having access to all of the Data Sources Wizard and the like.

We've been working hard on HDPI features, and every time we peel a layer off of the onion so to speak. We've been able to add some fundamental features, including the ability for MDI Child forms to know that DPI has changed. In .NET 7 we're devoting more resources to try to peel enough layers in the onion to allow most people to start taking advantages of it.

**Reply by cadilhac** [reply]
Thanks for the answer. But no words on components licensing...

**Reply by kirsan31** [reply]
> Thanks for the answer. But no words on components licensing...

I may be wrong, but this will be part of SDK...

**Comment by kirsan31** [comment]
The new article about designer has come out: https://devblogs.microsoft.com/dotnet/state-of-the-windows-forms-designer-for-net-applications/
And with preview of new [SDK](https://www.nuget.org/packages/Microsoft.WinForms.Designer.SDK) I was able partially [port](https://github.com/kirsan31/winforms-datavisualization/commit/17698a2255af0eb504588a6ee58d83cc36394924) [winforms-datavisualization](https://github.com/dotnet/winforms-datavisualization) to new designer.
All the remaining questions are with custom editors - looking forward to explanation how to port them...
At least the problem described [here](https://github.com/dotnet/winforms/discussions/5220#discussioncomment-987739) is solved.
